### PR TITLE
Update for Marco's comments #5 + new examples

### PR DESCRIPTION
**Pull request by mrdeep1**

### CODE REVIEW

File: draft-ietf-core-new-block.xml [hunk 1]
```diff
@@ -303,12 +303,16 @@
...
         <t>Q-Block2 Option is useful with GET, POST, PUT, FETCH, PATCH, and
-        iPATCH requests and their payload-bearing responses (2.01, 2.03, 2.04,
-        and 2.05) (Section 5.5 of <xref target="RFC7252"></xref>).</t>
+        iPATCH requests and their payload-bearing responses (2.01, 2.02, 2.03,
+        2.04, and 2.05) (Section 5.5 of <xref target="RFC7252"></xref>).</t>
+
+        <t>However, with methods needing both Q-Block1 for the request and
+        Q-Block2 for the response, and there is need for recovery following
+        payload loss, the numbers of packets needed to initiate and complete
+        the recovery can get large.</t>
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        the recovery can get very large as a function of the severity of the experienced loss.</t>
```

File: draft-ietf-core-new-block.xml [hunk 2]
```diff
@@ -668,7 +672,24 @@
         the Size1 Option.</t>
       </section>
 
-      <section title="Using Observe and Q-Block2 Options">
+      <section title="Using Observe Request and Q-Block1 Options">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
      <section title="Using Observe and Q-Block1 Options">
```

File: draft-ietf-core-new-block.xml [hunk 3]
```diff
@@ -668,7 +672,24 @@
...
+        <t>As the blocks of the body are sent without waiting for
+        acknowledgement of the individual blocks, the Observe value <xref
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        acknowledgement of the individual blocks, the observe value <xref
```

File: draft-ietf-core-new-block.xml [hunk 4]
```diff
@@ -668,7 +672,24 @@
...
+        target="RFC7641"></xref> MUST be the same for all the blocks of the
+        same body.</t>
+
+        <t>If the server reports any missing payloads, the client re-sends the
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <t>If the server reports any missing payload, the client re-sends the
```

File: draft-ietf-core-new-block.xml [hunk 5]
```diff
@@ -668,7 +672,24 @@
...
+        missing payload(s). Each re-sent payload is treated as a new Request
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        missing payload(s). Each re-sent payload is treated as a new request
```

File: draft-ietf-core-new-block.xml [hunk 6]
```diff
@@ -668,7 +672,24 @@
...
+        and the Observe Option MUST be included in the request.</t>
+
+        <t>If the response (or associated response) uses Q-Block2, and one of
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <t>If the response (or associated response) uses Q-Block2 Option, and one of
```

File: draft-ietf-core-new-block.xml [hunk 7]
```diff
@@ -668,7 +672,24 @@
...
+        the response payloads is missing, the request for the missing
+        payload(s) is treated as a new request. The request MUST comprise of
+        all of the request payloads and the Observe Option MUST NOT be
+        included in either the request or response.</t>
+      </section>
+
+      <section title="Using Observe Response and Q-Block2 Options">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
      <section title="Using Observe and Q-Block2 Options">
```

File: draft-ietf-core-new-block.xml [hunk 8]
```diff
@@ -850,6 +871,12 @@ missing-block-number = uint
         payload before requesting retransmission. NON_RECEIVE_TIMEOUT has a
         value of twice NON_TIMEOUT.</t>
 
+        <t>NON_MAX_RETRANSMIT is the maximum number of times a request for the
+        retransmission of missings payloads can occur without a response from
+        the remote peer. After this occurs, the peer SHOULD consider the body
+        stale and remove all references to it.  NON_MAX_RETRANSMIT has the
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        stale and remove all references to it.  By default, NON_MAX_RETRANSMIT has the
```

File: draft-ietf-core-new-block.xml [hunk 9]
```diff
@@ -1050,7 +1080,7 @@ Table 3: Congestion Control Parameters]]></artwork>
       to be followed.</t>
     </section>
 
-    <section title="Examples of Selective Block Recovery">
+    <section title="Examples using Q-Block1 and / or Q-Block2">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
    <section title="Examples using Q-Block1 and Q-Block2 Options">
```

File: draft-ietf-core-new-block.xml [hunk 10]
```diff
@@ -1073,13 +1103,14 @@ X<--: Message loss (response)
...
 
       <section title="Q-Block1 Option: Non-Confirmable Examples">
-        <t><xref target="B3non"></xref> depicts an example of a NON PUT
-        request conveying Q-Block1 Option. All the blocks are received by the
-        server.</t>
-
-        <t><figure anchor="B3non"
-            title="Example of NON Request with Q-Block1 Option (Without Loss)">
-            <artwork><![CDATA[        CoAP        CoAP
+        <section title="Q-Block1 - Simple Example">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="A Simple Example">
```

File: draft-ietf-core-new-block.xml [hunk 11]
```diff
@@ -1089,15 +1120,17 @@ X<--: Message loss (response)
...
          |   ...    |
 ]]></artwork>
-          </figure></t>
+            </figure></t>
 
-        <t><xref target="B3non0"></xref> depicts an example of a NON PUT
-        request conveying Q-Block1 Option. The number of payloads exceeds
-        MAX_PAYLOADS. All the blocks are received by the server.</t>
+        </section>
+        <section title="Q-Block1 - Handling MAX_PAYLOAD Limits">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="Handling MAX_PAYLOADS Limits">
```

File: draft-ietf-core-new-block.xml [hunk 12]
```diff
@@ -1111,15 +1144,17 @@ X<--: Message loss (response)
...
          |   ...    |
 ]]></artwork>
-          </figure></t>
+            </figure></t>
 
-        <t>Consider now a scenario where a new body of data is to be sent by
-        the client, but some blocks are dropped in transmission as illustrated
-        in <xref target="B3non1"></xref>.</t>
+        </section>
+        <section title="Q-Block1 - Handling MAX_PAYLOAD with Recovery">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="Handling MAX_PAYLOADS with Recovery">
```

File: draft-ietf-core-new-block.xml [hunk 13]
```diff
@@ -1164,27 +1199,29 @@ X<--: Message loss (response)
...
-        Options. The Q-Block2 Option indicates a block size hint (1024 bytes).
-        This request is replied to by the server using four (4) blocks that
-        are transmitted to the client without any loss. Each of these blocks
-        carries a Q-Block2 Option. The same process is repeated when an
-        Observe is triggered, but no loss is experienced by any of the
-        notification blocks.</t>
-
-        <t><figure anchor="nonb4"
-            title="Example of NON Notifications with Q-Block2 Option (Without Loss)">
+        <section title="Q-Block2 - Simple Example">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="A Simple Example">
```

File: draft-ietf-core-new-block.xml [hunk 14]
```diff
@@ -1164,27 +1199,29 @@ X<--: Message loss (response)
...
+          <t><xref target="nonb4"></xref> illustrates the example of Q-Block2
+          Option. The client sends a NON GET carrying an Observe and a Q-Block2
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
          Option. The client sends a NON GET carrying Observe and Q-Block2
```

File: draft-ietf-core-new-block.xml [hunk 15]
```diff
@@ -1204,12 +1241,14 @@ X<--: Message loss (response)
 ]]></artwork>
           </figure></t>
 
-        <t><xref target="nonb40"></xref> illustrates the same as <xref
-        target="nonb4"></xref> but this time has eleven (11) payloads which
-        exceeds MAX_PAYLOADS. There is no loss experienced.</t>
+        </section>
+        <section title="Q-Block2 - Handling MAX_PAYLOAD Limits">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="Handling MAX_PAYLOADS Limits">
```

File: draft-ietf-core-new-block.xml [hunk 16]
```diff
@@ -1240,14 +1279,16 @@ X<--: Message loss (response)
 ]]></artwork>
           </figure></t>
 
-        <t><xref target="nonb41"></xref> shows the example of an Observe that
-        is triggered but for which some notification blocks are lost. The
-        client detects the missing blocks and requests their retransmission.
-        It does so by indicating the blocks that were successfully
-        received.</t>
+        </section>
+        <section title="Q-Block2 - Handling MAX_PAYLOAD with Recovery">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="Handling MAX_PAYLOADS with Recovery">
```

File: draft-ietf-core-new-block.xml [hunk 17]
```diff
@@ -1280,17 +1321,19 @@ X<--: Message loss (response)
...
+          <t>Under high levels of traffic loss, the client can elect not to
+          retry getting missing blocks of data. This decision is implementation
+          specific.</t>
 
-        <t><xref target="nonb411"></xref> shows the example of an Observe that
-        is triggered but only the first two notification blocks reaches the
-        client. In order to retrieve the missing blocks, the client sends a
-        request with a single Q-Block2 Option with the M bit set.</t>
+        </section>
+        <section title="Q-Block2 - Handling Recovery using M bit set">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="Handling Recovery using M-bit Set">
```

File: draft-ietf-core-new-block.xml [hunk 18]
```diff
@@ -1280,17 +1321,19 @@ X<--: Message loss (response)
...
+          <t><xref target="nonb411"></xref> shows the example of an Observe that
+          is triggered but only the first two notification blocks reaches the
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
          is triggered but only the first two notification blocks reach the
```

File: draft-ietf-core-new-block.xml [hunk 19]
```diff
@@ -1323,88 +1366,179 @@ X<--: Message loss (response)
...
 ]]></artwork>
           </figure></t>
+        </section>
       </section>
 
       <section title="Q-Block1 and Q-Block2 Options: Non-Confirmable Examples">
-        <t><xref target="b12non"></xref> illustrates the example of a FETCH
-        using both Q-Block1 and Q-Block2 along with an Observe. No loss is
-        experienced.</t>
+        <section title="Q-Block + Q-Block2 - Simple Example">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="A Simple Example">
```

File: draft-ietf-core-new-block.xml [hunk 20]
```diff
@@ -1323,88 +1366,179 @@ X<--: Message loss (response)
...
+          <t><xref target="b12non"></xref> illustrates the example of a FETCH
+          using both Q-Block1 and Q-Block2 along with an Observe. No loss is
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
          using both Q-Block1 and Q-Block2 Options along with an Observe Option. No loss is
```

File: draft-ietf-core-new-block.xml [hunk 21]
```diff
@@ -1323,88 +1366,179 @@ X<--: Message loss (response)
...
 
 ]]></artwork>
           </figure></t>
 
-        <t><xref target="b12non0"></xref> illustrates the same as <xref
-        target="b12non"></xref> but this time has eleven (11) payloads in both
-        directions which exceeds MAX_PAYLOADS. There is no loss
-        experienced.</t>
+        </section>
+        <section title="Q-Block1 + Q-Block2 - Handling MAX_PAYLOAD Limits">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="Handling MAX_PAYLOADS Limits">
```

File: draft-ietf-core-new-block.xml [hunk 22]
```diff
@@ -1323,88 +1366,179 @@ X<--: Message loss (response)
...
 
 ]]></artwork>
           </figure></t>
 
-        <t>If there is a loss of a message with a Q-Block1 Option, this is
-        handled as shown in <xref target="B3non1"></xref> and <xref
-        target="B3non2"></xref> with FETCH substitued for PUT and 2.04 becomes
-        a 2.05 response.</t>
+        </section>
+        <section title="Q-Block1 + Q-Block2 - Handling Recovery">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <section title="Handling Recovery">
```

File: draft-ietf-core-new-block.xml [hunk 23]
```diff
@@ -1323,88 +1366,179 @@ X<--: Message loss (response)
...
+         |   ...    |
+      [[NON_RECEIVE_TIMEOUT (server) delay expires]]
+]]></artwork>
+          </figure></t>
+
+          <t>The server realizes that some blocks are missing and asks for the
+          missing blocks in one go (<xref
+          target="b12non2"></xref>). It does so by indicating which blocks have
+          been received in the data portion of the response. The token used in
+          the response should be the token that was used in the last block
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
          the response is the token that was used in the last block
```

File: draft-ietf-core-new-block.xml [hunk 24]
```diff
@@ -1134,20 +1169,20 @@ X<--: Message loss (response)
...
-
-        <t><figure anchor="B3non2"
-            title="Example of NON Request with Q-Block1 Option (Blocks Recovery)">
-            <artwork><![CDATA[        CoAP        CoAP
+            </figure></t>
+
+          <t>On seeing a payload from the next set of payloads, the server
+          realizes that some blocks are missing from the previous MAX_PAYLOADS
+          payloads and asks for the missing blocks in one go (<xref
+          target="B3non2"></xref>). It does so by indicating which blocks have
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml:1178]: Why are we deviating from reporting missing block to ACking received ones?

**Review by boucadair — CHANGES_REQUESTED** [review]

File: draft-ietf-core-new-block.xml (hunk 1 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 2 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 3 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 4 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 5 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 6 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 7 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: UPdated

File: draft-ietf-core-new-block.xml (hunk 8 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 9 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Updated

File: draft-ietf-core-new-block.xml (hunk 24 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml:1178]: Well caught - now have `blocks have not been received` - fixed in 2 places.

**Comment by mrdeep1** [comment]
Updates pushed

**Comment by boucadair** [comment]
Thanks. I see some minor edits not taken into account (e.g., all the titles with "MAX_PAYLOAD"). 

I will merge and fix these in the master directly. I will be doing a last check and submit.

**Comment by mrdeep1** [comment]
Apologies - missed your plural change to MAX_PAYLOAD.

**Comment by boucadair** [comment]
See all the fixes at 3d9419cc8f9bdacbaa192561fa09a59a2ce3b19b

**Comment by mrdeep1** [comment]
Fixes in https://github.com/core-wg/new-block/commit/3d9419cc8f9bdacbaa192561fa09a59a2ce3b19b look good to me.

**Pull request merged**